File: python/structures/src/sorts/bubble_sort.py
from structures.src.util import sort_util

def sort(arr):
    # want this to be pure
    out = list(arr)

    # keep doing it until we're sorted
    while !sort_util.is_sorted(out):
        for i in range(0, len(out) - 1):
            # if this is out of order, swap it to the next slot
            if out[i] > out[i + 1]:
                swap(out, i, i + 1)

    return out

def swap(arr, i, j):
    tmp = arr[i]
    arr[i] = arr[j]
    arr[j] = tmp

File: Scala/main/scala/com/ckeyser/sorts/BubbleSort.scala
package com.ckeyser.sorts

import Ordering.Implicits._
import com.ckeyser.util.SortUtil

/**
  * Created by Chris Keyser on 9/28/2016.
  */
object BubbleSort {
    def sort[A : Ordering](arr: Array[A]): Array[A] = {
        val out = arr.clone()

        while(!SortUtil.isSorted(out))
            for(i <- Range(0, out.length - 1))
              if(arr(i) > arr(i + 1))
                SortUtil.swap(arr, i, i + 1)

        arr
    }
}
